File: Documentation/5.4/Samples/python/ClientApi/Operations/Maintenance/Indexes/GetIndexErrors.py
import datetime
from typing import Optional, List

from ravendb import GetIndexErrorsOperation
from ravendb.documents.indexes.definitions import IndexingError, IndexErrors
from ravendb.documents.operations.definitions import MaintenanceOperation

from examples_base import ExampleBase


class GetIndexErrors(ExampleBase):
    def setUp(self):
        super().setUp()

    def test_get_index_errors(self):
        with self.embedded_server.get_document_store("GetIndexErrors") as store:
            # region get_errors_all
            # Define the get index errors operation
            get_index_errors_op = GetIndexErrorsOperation()

            # Execute the operation by passing it to maintenance.send
            index_errors = store.maintenance.send(get_index_errors_op)

            # index_errors will contain errors for ALL indexes
            # endregion

            # region get_errors_specific
            # Define the get index errors operation for specific indexes
            get_index_errors_op = GetIndexErrorsOperation("Orders/Totals")

            # Execute the operation by passing it to maintenance.send
            # An exception will be thrown if any of the specified indexes do not exist
            index_errors = store.maintenance.send(get_index_errors_op)

            # index_errors will contain errors only for index "Orders/Totals"
            # endregion

    class Foo:
        # region syntax_1
        class GetIndexErrorsOperation(MaintenanceOperation[List[IndexErrors]]):
            def __init__(self, *index_names: str):  # If no index_names provided, get errors for all indexes
                ...

        # endregion
        # region syntax_2
        class IndexErrors:
            def __init__(self, name: Optional[str] = None, errors: Optional[List[IndexingError]] = None):
                self.name = name  # Index name
                self.errors = errors  # List of errors for this index

        # endregion

        # region syntax_3
        class IndexingError:
            def __init__(
                self,
                error: Optional[str] = None,
                timestamp: Optional[datetime.datetime] = None,
                document: Optional[str] = None,
                action: Optional[str] = None,
            ):
                # Error message
                self.error = error

                # Time of error
                self.timestamp = timestamp

                # If action is 'Map'    - field will contain the document ID
                # If action is 'Reduce' - field will contain the Reduce key value
                # For all other actions - field will be None
                self.document = document

                # Area where error has occurred, e.g. Map/Reduce/Analyzer/Memory/etc.
                self.action = action

        # endregion

File: Documentation/5.4/Samples/php/ClientApi/Operations/Maintenance/Indexes/GetIndexErrors.php
<?php

namespace RavenDB\Samples\ClientApi\Operations\Maintenance\Indexes;

use RavenDB\Documents\DocumentStore;
use RavenDB\Documents\Indexes\IndexErrorsArray;
use RavenDB\Documents\Operations\Indexes\GetIndexErrorsOperation;

class GetIndexErrors
{
    public function samples(): void
    {
        $store = new DocumentStore();
        try {
            # region get_errors_all
            // Define the get index errors operation
            $getIndexErrorsOp = new GetIndexErrorsOperation();

            // Execute the operation by passing it to maintenance.send
            /** @var IndexErrorsArray $indexErrors */
            $indexErrors = $store->maintenance()->send($getIndexErrorsOp);

            // indexErrors will contain errors for ALL indexes
            # endregion
        } finally {
            $store->close();
        }

        $store = new DocumentStore();
        try {
            # region get_errors_specific
            // Define the get index errors operation for specific indexes
            $getIndexErrorsOp = new GetIndexErrorsOperation([ "Orders/Totals" ]);

            // Execute the operation by passing it to Maintenance.Send
            // An exception will be thrown if any of the specified indexes do not exist
            /** @var IndexErrorsArray $indexErrors */
            $indexErrors = $store->maintenance()->send($getIndexErrorsOp);

            // indexErrors will contain errors only for index "Orders/Totals"
            # endregion
        } finally {
            $store->close();
        }
    }
}

/*
interface IFoo
{
    # region syntax_1
    // Available overloads:
    GetIndexErrorsOperation()                  // Get errors for all indexes
    GetIndexErrorsOperation(array $indexNames) // Get errors for specific indexes
    # endregion
}
*/

/*
# region syntax_2
public class IndexErrors
{
    private ?string $name = null;               // Index name
    private ?IndexingErrorArray $errors = null; // List of errors for this index

    // ... getters and setters
}
# endregion
*/

/*
# region syntax_3
public class IndexingError
{
    // The error message
    private ?string $error = null;

    // Time of error
    private ?DateTimeInterface $timestamp = null;

    // If Action is 'Map'    - field will contain the document ID
    // If Action is 'Reduce' - field will contain the Reduce key value
    // For all other Actions - field will be null
    private ?string $document = null;

    // Area where error has occurred, e.g. Map/Reduce/Analyzer/Memory/etc.
    private ?string $action = null;

    // ... getters and setters
}
# endregion
*/
